server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -377,19 +393,24 @@ interface CacheEntity extends Accountable {
      */
     static class Key implements Accountable, Writeable {
         public final ShardId shardId; // use as identity equality
+        public final int indexShardHashCode; // While ShardId is usually sufficient to uniquely identify an
+        // indexShard but in case where the same indexShard is deleted and reallocated on same node, we need the
+        // hashcode(default) to identify the older indexShard but with same shardId.
         public final String readerCacheKeyId;
         public final BytesReference value;
 
-        Key(ShardId shardId, BytesReference value, String readerCacheKeyId) {
+        Key(ShardId shardId, BytesReference value, String readerCacheKeyId, int indexShardHashCode) {
             this.shardId = shardId;
             this.value = value;
             this.readerCacheKeyId = Objects.requireNonNull(readerCacheKeyId);
+            this.indexShardHashCode = indexShardHashCode;
```

Comment:
If this is intended to be the hashcode of the `ShardId`, couldn't we just set it as `shardId.hashCode()`?

Oh -- is this the hashCode of the `IndexShard` object, not the `ShardId`? 

If the goal is to associate the key with the specific instance of the `IndexShard`, you should probably use `System.identityHashCode` instead.

Reply:
Yeah this is the hashcode of `IndexShard` object. Using shardId hashcode will not work with the same reason.
I will change it to `System.identityHashCode`